// @elastic/eui/lib/services/format/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _format_auto = require('./format_auto');

Object.defineProperty(exports, 'formatAuto', {
  enumerable: true,
  get: function get() {
    return _format_auto.formatAuto;
  }
});

var _format_boolean = require('./format_boolean');

Object.defineProperty(exports, 'formatBoolean', {
  enumerable: true,
  get: function get() {
    return _format_boolean.formatBoolean;
  }
});

var _format_date = require('./format_date');

Object.defineProperty(exports, 'formatDate', {
  enumerable: true,
  get: function get() {
    return _format_date.formatDate;
  }
});
Object.defineProperty(exports, 'dateFormatAliases', {
  enumerable: true,
  get: function get() {
    return _format_date.dateFormatAliases;
  }
});

var _format_number = require('./format_number');

Object.defineProperty(exports, 'formatNumber', {
  enumerable: true,
  get: function get() {
    return _format_number.formatNumber;
  }
});

var _format_text = require('./format_text');

Object.defineProperty(exports, 'formatText', {
  enumerable: true,
  get: function get() {
    return _format_text.formatText;
  }
});
