// @elastic/eui/lib/services/color/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _is_color_dark = require('./is_color_dark');

Object.defineProperty(exports, 'isColorDark', {
  enumerable: true,
  get: function get() {
    return _is_color_dark.isColorDark;
  }
});

var _hex_to_rgb = require('./hex_to_rgb');

Object.defineProperty(exports, 'hexToRgb', {
  enumerable: true,
  get: function get() {
    return _hex_to_rgb.hexToRgb;
  }
});

var _rgb_to_hex = require('./rgb_to_hex');

Object.defineProperty(exports, 'rgbToHex', {
  enumerable: true,
  get: function get() {
    return _rgb_to_hex.rgbToHex;
  }
});

var _luminance_and_contrast = require('./luminance_and_contrast');

Object.defineProperty(exports, 'calculateContrast', {
  enumerable: true,
  get: function get() {
    return _luminance_and_contrast.calculateContrast;
  }
});
Object.defineProperty(exports, 'calculateLuminance', {
  enumerable: true,
  get: function get() {
    return _luminance_and_contrast.calculateLuminance;
  }
});

var _visualization_colors = require('./visualization_colors');

Object.defineProperty(exports, 'VISUALIZATION_COLORS', {
  enumerable: true,
  get: function get() {
    return _visualization_colors.VISUALIZATION_COLORS;
  }
});
Object.defineProperty(exports, 'DEFAULT_VISUALIZATION_COLOR', {
  enumerable: true,
  get: function get() {
    return _visualization_colors.DEFAULT_VISUALIZATION_COLOR;
  }
});
